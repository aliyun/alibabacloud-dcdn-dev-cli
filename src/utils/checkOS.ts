import os from 'os';

export enum Platforms {
  Win = 'win',
  AppleArm = 'darwin-arm64',
  AppleIntel = 'darwin-x86_64',
  Linux = 'linux',
  others = 'others'
}

export const checkOS = (): Platforms => {
  const platform = os.platform();
  const cpus = os.cpus();
  const cpuModel = cpus[0].model.toLowerCase();
  if (platform === 'win32') {
    return Platforms.Win;
  }
  if (platform === 'darwin') {
    if (cpuModel.includes('apple')) {
      return Platforms.AppleArm;
    } else {
      return Platforms.AppleIntel;
    }
  }
  if (platform === 'linux') {
    return Platforms.Linux;
  }
  return Platforms.others;
};
