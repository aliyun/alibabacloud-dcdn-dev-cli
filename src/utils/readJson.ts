
import { readFileSync } from 'fs'

export const readJson = (jsonPath: string) => {
    try {
        const json = JSON.parse(readFileSync(jsonPath, 'utf8'));
        return json;
    } catch (_) {
        throw new TypeError(`${jsonPath} is not a valid JSON file.`);
    }
}