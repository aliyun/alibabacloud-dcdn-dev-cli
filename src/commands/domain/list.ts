import { CommandModule, ArgumentsCamelCase, Argv } from 'yargs';
import { checkDirectory, checkIsLoginSuccess } from '../utils.js';
import logger from '../../libs/logger.js';
import { getProjectConfig } from '../../utils/fileUtils/index.js';
import { GetRoutineReq, RelatedRecordProps } from '../../libs/interface.js';
import { ApiService } from '../../libs/apiService.js';
import t from '../../i18n/index.js';
import { validRoutine } from '../../utils/checkIsRoutineCreated.js';

const listDomain: CommandModule = {
  command: 'list',
  describe: `🔍 ${t('domain_list_describe').d('List all related domains')}`,
  handler: async (argv: ArgumentsCamelCase) => {
    handleListDomains(argv);
  }
};

export default listDomain;

export async function handleListDomains(argv: ArgumentsCamelCase) {
  if (!checkDirectory()) {
    return;
  }

  const projectConfig = getProjectConfig();
  if (!projectConfig) return logger.notInProject();

  const isSuccess = await checkIsLoginSuccess();
  if (!isSuccess) return;

  await validRoutine(projectConfig.name);

  const server = await ApiService.getInstance();

  const req: GetRoutineReq = { Name: projectConfig.name };
  const routineDetail = await server.getRoutine(req);

  if (!routineDetail) return;
  const relatedRecords: RelatedRecordProps[] =
    routineDetail.data?.RelatedRecords ?? [];

  if (relatedRecords.length === 0) {
    logger.log(`🙅 ${t('domain_list_empty').d('No related domains found')}`);
    return;
  }
  const domainList = relatedRecords.map((record) => record.RecordName);
  logger.log(`📃 ${t('domain_list_title').d('Related domains:')}`);
  logger.tree(domainList);
}
