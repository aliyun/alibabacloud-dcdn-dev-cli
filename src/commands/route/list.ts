import { CommandModule, ArgumentsCamelCase, Argv } from 'yargs';
import { getProjectConfig } from '../../utils/fileUtils/index.js';
import { GetRoutineReq, RelatedRouteProps } from '../../libs/interface.js';
import Table from 'cli-table3';
import { checkDirectory, checkIsLoginSuccess } from '../utils.js';
import { ApiService } from '../../libs/apiService.js';
import logger from '../../libs/logger.js';
import t from '../../i18n/index.js';
import { validRoutine } from '../../utils/checkIsRoutineCreated.js';

const listRoute: CommandModule = {
  command: 'list',
  describe: `🔍 ${t('route_list_describe').d('List all related routes')}`,
  handler: async () => {
    handleListRoutes();
  }
};

export default listRoute;

export async function handleListRoutes() {
  if (!checkDirectory()) {
    return;
  }

  const projectConfig = getProjectConfig();
  if (!projectConfig) return logger.notInProject();
  const isSuccess = await checkIsLoginSuccess();
  if (!isSuccess) return;

  await validRoutine(projectConfig.name);

  const server = await ApiService.getInstance();
  const req: GetRoutineReq = { Name: projectConfig.name };

  const routineDetail = await server.getRoutine(req);
  if (!routineDetail) return;
  const relatedRoutes: RelatedRouteProps[] =
    routineDetail.data?.RelatedRoutes ?? [];

  if (relatedRoutes.length === 0) {
    logger.warn(`🙅 ${t('route_list_empty').d('No related routes found')}`);
    return;
  }
  logger.log(`📃 ${t('route_list_title').d('Related routes')}:`);
  displayRelatedRouteList(relatedRoutes);
}

export async function displayRelatedRouteList(routeList: RelatedRouteProps[]) {
  const table = new Table({
    head: ['Route', 'Site'],
    colWidths: [30, 30]
  });
  for (let i = 0; i < routeList.length; i++) {
    const route = routeList[i];
    table.push([route.Route, route.SiteName]);
  }
  console.log(table.toString());
}
