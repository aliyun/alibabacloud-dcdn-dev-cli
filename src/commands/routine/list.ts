import { CommandModule, ArgumentsCamelCase, Argv } from 'yargs';
import { EdgeFunctionItem, ListSitesReq } from '../../libs/interface.js';
import Table from 'cli-table3';
import logger from '../../libs/logger.js';
import { Base64 } from 'js-base64';
import { checkIsLoginSuccess } from '../utils.js';
import chalk from 'chalk';
import { ApiService } from '../../libs/apiService.js';
import t from '../../i18n/index.js';

const list: CommandModule = {
  command: 'list',
  describe: `ðŸ“‹ ${t('list_describe').d('List all your routines')}`,
  builder: (yargs: Argv) => {
    return (
      yargs
        // .positional('deployment', {
        //   type: 'string',
        //   description: t('list_deployment_positional_describe').d(
        //     '(Optional) ID or URL of the deployment to tail. Specify by environment if deployment ID is unknown.'
        //   ),
        //   demandOption: true
        // })
        // .option('site', {
        //   alias: 's',
        //   describe: t('list_site_option_describe').d(
        //     'List all site names of your account'
        //   ),
        //   type: 'boolean',
        //   default: false
        // })
        .usage(`${t('common_usage').d('Usage')}: \$0 list []`)
    );
  },
  handler: async (argv: ArgumentsCamelCase) => {
    handleList(argv);
  }
};

export default list;

export async function handleList(argv: ArgumentsCamelCase) {
  const { site, ...args } = argv;
  const isSuccess = await checkIsLoginSuccess();
  if (!isSuccess) return;
  const server = await ApiService.getInstance();

  if (site) {
    const req: ListSitesReq = {
      SiteSearchType: 'fuzzy',
      Status: 'active',
      PageNumber: 1,
      PageSize: 50
    };
    const res = await server.listSites(req);
    const siteList = res?.data.Sites ?? [];
    const siteNameList: string[] = siteList?.map((item: any) => item.SiteName);
    logger.log(
      chalk.bold.bgGray(
        `ðŸ“ƒ ${t('list_site_name_title').d('List all of site names')}:`
      )
    );
    logger.tree(siteNameList);
    return;
  }

  const res = await server.getRoutineUserInfo();
  const routineList = res?.Routines;
  if (routineList) {
    console.log(
      chalk.bold.bgGray(
        `ðŸ“ƒ ${t('list_routine_name_title').d('List all of routine')}:`
      )
    );
    displayRoutineList(routineList);
  }
}

export async function displayRoutineList(versionList: EdgeFunctionItem[]) {
  const table = new Table({
    head: ['Name', 'Created', 'Description', 'Specification']
  });

  versionList.forEach((version) => {
    table.push([
      version.RoutineName,
      new Date(version.CreateTime).toLocaleString(),
      Base64.decode(version.Description),
      version.SpecName ?? ' '
    ]);
  });

  console.log(table.toString());
}
