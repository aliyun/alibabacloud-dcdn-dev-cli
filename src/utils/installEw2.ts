import os from 'os';
import fs from 'fs-extra';
import path from 'path';
import { execSync } from 'child_process';
import { checkOS, Platforms } from './checkOS.js';
import logger from '../libs/logger.js';
import { getDirName } from './fileUtils/base.js';
import t from '../i18n/index.js';

export const EW2DirName = '.ew2';
export const EW2BinName = 'edgeworker2';
export const EW2Path = path.join(os.homedir(), EW2DirName);
export const EW2BinPath = path.join(EW2Path, EW2BinName);

export function preCheckEw2(): Promise<boolean> {
  return new Promise((resolve, reject) => {
    fs.access(EW2BinPath, fs.constants.X_OK, (err) => {
      if (err) {
        if (err.code === 'EACCES') {
          logger.pathEacces(EW2BinPath);
          reject(false);
          return;
        }
        if (err.code === 'ENOENT' || err.code === 'EPERM') {
          installEw2()
            .then(() => {
              resolve(true);
            })
            .catch((err) => {
              console.log('err', err);
              reject(false);
            });
        }
      } else {
        resolve(true);
      }
    });
  });
}
export async function installEw2(): Promise<boolean> {
  const __dirname = getDirName(import.meta.url);
  const p = path.join(__dirname, './install');
  const installCommand = `sh ${p}/installEw2.sh`;
  logger.warn(
    t('install_runtime_tip').d(
      `ðŸ”” Runtime must be installed to use esa dev. Installing...`
    )
  );
  try {
    execSync(installCommand, { stdio: 'inherit' });
    logger.success(t('install_runtime_success').d(`Runtime installed.`));
    return true;
  } catch (error) {
    logger.error(`Failed to install: ${(error as Error).message}`);
    return false;
  }
}
