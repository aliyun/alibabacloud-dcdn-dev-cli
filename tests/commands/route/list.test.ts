import { it, describe, expect, vi } from 'vitest';
import { handleListRoutes } from '../../../src/commands/route/list.js';
import { validDomain, validName } from '../../../src/commands/utils.js';
import { mockConsoleMethods } from '../../helper/mockConsole.js';
import { ApiService } from '../../../src/libs/apiService.js';
import logger from '../../../src/libs/logger.js';

describe('handle list related routes', () => {
  let std = mockConsoleMethods();
  vi.spyOn(logger, 'error').mockImplementation(() => {});
  vi.spyOn(logger, 'warn').mockImplementation(() => {});

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should handle displaying related routes success', async () => {
    await handleListRoutes();
    expect(std.out).toMatchInlineSnapshot(`
      [MockFunction log] {
        "calls": [
          [
            "📃 Related routes:",
          ],
          [
            "┌──────────────────────────────┬──────────────────────────────┐
      │ Route                        │ Site                         │
      ├──────────────────────────────┼──────────────────────────────┤
      │ test.com/1                   │ test.com                     │
      ├──────────────────────────────┼──────────────────────────────┤
      │ test.com/2                   │ test.com                     │
      └──────────────────────────────┴──────────────────────────────┘",
          ],
        ],
        "results": [
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    `);
  });

  it('should handle no related routes found', async () => {
    vi.mocked((await ApiService.getInstance()).getRoutine).mockResolvedValue({
      data: {
        RelatedRoutes: []
      }
    } as any);
    await handleListRoutes();
    expect(logger.warn).toBeCalledWith('🙅 No related routes found');
  });
});
