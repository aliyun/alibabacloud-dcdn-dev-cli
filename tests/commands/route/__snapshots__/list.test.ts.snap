// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handle list related routes > should handle displaying related routes success 1`] = `
[MockFunction log] {
  "calls": [
    [
      "📃 Related routes:",
    ],
    [
      "[90m┌────────────[39m[90m┬──────────┐[39m
[90m│[39m[31m Route      [39m[90m│[39m[31m Site     [39m[90m│[39m
[90m├────────────[39m[90m┼──────────┤[39m
[90m│[39m test.com/1 [90m│[39m test.com [90m│[39m
[90m├────────────[39m[90m┼──────────┤[39m
[90m│[39m test.com/2 [90m│[39m test.com [90m│[39m
[90m└────────────[39m[90m┴──────────┘[39m",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
